rust/rope/src/tree.rs

```diff
@@ -763,25 +713,35 @@ impl<'a, N: NodeInfo> Cursor<'a, N> {
         CursorIter { cursor: self, _metric: PhantomData }
     }
 
+    /// Tries to find the last boundary in the leaf the cursor is currently in.
+    ///
+    /// If the last boundary is at the end of the leaf, it is only counted if
+    /// it is less than `orig_pos`.
+    #[inline]
+    fn last_inside_leaf<M: Metric<N>>(&mut self, orig_pos: usize) -> Option<usize> {
+        let l = self.leaf.expect("inconsistent, shouldn't get here");
+        let len = l.len();
+        if self.offset_of_leaf + len < orig_pos && M::is_boundary(l, len) {
+            let _ = self.next_leaf();
+            return Some(self.position);
+        }
+        let offset_in_leaf = M::prev(l, len)?;
+        self.position = self.offset_of_leaf + offset_in_leaf;
+        Some(self.position)
+    }
+
     /// Tries to find the next boundary in the leaf the cursor is currently in.
     #[inline]
     fn next_inside_leaf<M: Metric<N>>(&mut self) -> Option<usize> {
-        if let Some(l) = self.leaf {
-            let offset_in_leaf = self.position - self.offset_of_leaf;
-            if let Some(offset_in_leaf) = M::next(l, offset_in_leaf) {
-                if offset_in_leaf == l.len()
-                    && self.offset_of_leaf + offset_in_leaf != self.root.len()
-                {
-                    let _ = self.next_leaf();
-                } else {
-                    self.position = self.offset_of_leaf + offset_in_leaf;
-                }
-                return Some(self.position);
-            }
+        let l = self.leaf.expect("inconsistent, shouldn't get here");
+        let offset_in_leaf = self.position - self.offset_of_leaf;
+        let offset_in_leaf = M::next(l, offset_in_leaf)?;
+        if offset_in_leaf == l.len() && self.offset_of_leaf + offset_in_leaf != self.root.len() {
+            let _ = self.next_leaf();
```

Comment:
okay, so maybe this answers my question: if we're at a boundary at the end of a leaf we always want to advance to the next leaf?

Reply:
Yes, the logic for this specific case is parallel.